File: src/harness/eval.py
"""Evaluation runner for batch-testing the agent harness against defined cases."""

from __future__ import annotations

import time
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List


@dataclass
class EvalCase:
    """A single evaluation case.

    Attributes:
        id: Unique identifier for the case.
        goal: The task goal string passed to the agent.
        context: Optional extra context dict.
        expected_stop_reason: If set, the run is considered passing only when
            ``RunResult.stop_reason`` matches this value.
        expected_keywords: If non-empty, the run is considered passing only when
            every keyword appears (case-insensitive) in the final response.
    """

    id: str
    goal: str
    context: Dict[str, Any] = field(default_factory=dict)
    expected_stop_reason: str | None = None
    expected_keywords: List[str] = field(default_factory=list)


@dataclass
class EvalCaseResult:
    """Result of running a single eval case.

    Attributes:
        id: Case identifier.
        passed: Whether all expectations were met.
        stop_reason: Actual stop reason from the run.
        turns: Number of turns used.
        final_response: Final response text.
        failures: List of failure reasons (empty when passed).
        duration_s: Wall-clock duration in seconds.
    """

    id: str
    passed: bool
    stop_reason: str
    turns: int
    final_response: str
    failures: List[str]
    duration_s: float


@dataclass
class EvalReport:
    """Aggregated results across all eval cases.

    Attributes:
        total: Total number of cases.
        passed: Number of cases that passed.
        failed: Number of cases that failed.
        pass_rate: Fraction of cases that passed (0.0–1.0).
        avg_turns: Average number of turns across all cases.
        avg_duration_s: Average wall-clock duration per case.
        results: Per-case results.
    """

    total: int
    passed: int
    failed: int
    pass_rate: float
    avg_turns: float
    avg_duration_s: float
    results: List[EvalCaseResult]

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to a plain dict suitable for JSON output."""
        return asdict(self)


class EvalRunner:
    """Runs a list of :class:`EvalCase` objects against a :class:`HarnessAgent`.

    Example::

        from harness import HarnessAgent, RunConfig, RuleBasedLLM
        from harness.eval import EvalCase, EvalRunner

        cases = [
            EvalCase(
                id="add_numbers",
                goal="please add numbers",
                expected_stop_reason="final_response",
                expected_keywords=["5"],
            ),
        ]
        agent = HarnessAgent(llm=RuleBasedLLM(), config=RunConfig())
        runner = EvalRunner(agent)
        report = runner.run(cases)
        print(report.pass_rate)
    """

    def __init__(self, agent: Any) -> None:
        """
        Args:
            agent: A :class:`~harness.agent.HarnessAgent` instance.
        """
        self.agent = agent

    def run(self, cases: List[EvalCase]) -> EvalReport:
        """Run all cases and return an aggregated :class:`EvalReport`.

        Args:
            cases: List of eval cases to execute.

        Returns:
            An :class:`EvalReport` with per-case results and aggregate metrics.
        """
        results: List[EvalCaseResult] = []
        for case in cases:
            result = self._run_case(case)
            results.append(result)

        total = len(results)
        passed = sum(1 for r in results if r.passed)
        avg_turns = sum(r.turns for r in results) / total if total else 0.0
        avg_duration = sum(r.duration_s for r in results) / total if total else 0.0

        return EvalReport(
            total=total,
            passed=passed,
            failed=total - passed,
            pass_rate=passed / total if total else 0.0,
            avg_turns=avg_turns,
            avg_duration_s=avg_duration,
            results=results,
        )

    def _run_case(self, case: EvalCase) -> EvalCaseResult:
        t0 = time.monotonic()
        run_result = self.agent.run(goal=case.goal, context=dict(case.context))
        duration = time.monotonic() - t0

        failures: List[str] = []

        if case.expected_stop_reason and run_result.stop_reason != case.expected_stop_reason:
            failures.append(
                f"stop_reason: expected={case.expected_stop_reason!r}, "
                f"actual={run_result.stop_reason!r}"
            )

        response_lower = run_result.final_response.lower()
        for kw in case.expected_keywords:
            if kw.lower() not in response_lower:
                failures.append(f"missing keyword in response: {kw!r}")

        return EvalCaseResult(
            id=case.id,
            passed=len(failures) == 0,
            stop_reason=run_result.stop_reason,
            turns=len(run_result.turns),
            final_response=run_result.final_response,
            failures=failures,
            duration_s=round(duration, 3),
        )

File: ts/src/eval.ts
/**
 * Evaluation runner for batch-testing the agent harness against defined cases.
 * Loads eval cases, runs HarnessAgent, compares output against expectations.
 */

import type { HarnessAgent } from "./agent.js";

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

export interface EvalCase {
  id: string;
  goal: string;
  context?: Record<string, unknown>;
  expected_stop_reason?: string;
  expected_keywords?: string[];
}

export interface EvalCaseResult {
  id: string;
  passed: boolean;
  stop_reason: string;
  turns: number;
  final_response: string;
  failures: string[];
  duration_s: number;
}

export interface EvalReport {
  total: number;
  passed: number;
  failed: number;
  pass_rate: number;
  avg_turns: number;
  avg_duration_s: number;
  results: EvalCaseResult[];
  config_version: string;
  toDict(): Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Built-in regression cases
// ---------------------------------------------------------------------------

export const BUILTIN_CASES: EvalCase[] = [
  {
    id: "add_numbers",
    goal: "please add numbers",
    expected_stop_reason: "final_response",
    expected_keywords: ["5"],
  },
  {
    id: "get_time",
    goal: "what is the current time",
    expected_stop_reason: "final_response",
    expected_keywords: [],
  },
  {
    id: "direct_answer",
    goal: "hello world",
    expected_stop_reason: "final_response",
    expected_keywords: [],
  },
];

// ---------------------------------------------------------------------------
// EvalRunner
// ---------------------------------------------------------------------------

export class EvalRunner {
  private agent: HarnessAgent;

  constructor(agent: HarnessAgent) {
    this.agent = agent;
  }

  /**
   * Run all cases and return an aggregated EvalReport.
   */
  async run(
    cases: EvalCase[],
    configVersion: string = "unversioned",
  ): Promise<EvalReport> {
    const results: EvalCaseResult[] = [];
    for (const c of cases) {
      const result = await this._runCase(c);
      results.push(result);
    }

    const total = results.length;
    const passed = results.filter((r) => r.passed).length;
    const avgTurns = total > 0 ? results.reduce((s, r) => s + r.turns, 0) / total : 0;
    const avgDuration =
      total > 0 ? results.reduce((s, r) => s + r.duration_s, 0) / total : 0;

    const report: EvalReport = {
      total,
      passed,
      failed: total - passed,
      pass_rate: total > 0 ? passed / total : 0,
      avg_turns: avgTurns,
      avg_duration_s: avgDuration,
      results,
      config_version: configVersion,
      toDict() {
        return {
          total: this.total,
          passed: this.passed,
          failed: this.failed,
          pass_rate: this.pass_rate,
          avg_turns: this.avg_turns,
          avg_duration_s: this.avg_duration_s,
          results: this.results.map((r) => ({ ...r })),
          config_version: this.config_version,
        };
      },
    };

    return report;
  }

  /**
   * Run a single eval case and return the result.
   */
  private async _runCase(evalCase: EvalCase): Promise<EvalCaseResult> {
    const t0 = performance.now();
    const runResult = await this.agent.run(
      evalCase.goal,
      evalCase.context ? { ...evalCase.context } : {},
    );
    const duration = (performance.now() - t0) / 1000;

    const failures: string[] = [];

    if (
      evalCase.expected_stop_reason &&
      runResult.stop_reason !== evalCase.expected_stop_reason
    ) {
      failures.push(
        `stop_reason: expected='${evalCase.expected_stop_reason}', ` +
          `actual='${runResult.stop_reason}'`,
      );
    }

    const responseLower = runResult.final_response.toLowerCase();
    const keywords = evalCase.expected_keywords ?? [];
    for (const kw of keywords) {
      if (!responseLower.includes(kw.toLowerCase())) {
        failures.push(`missing keyword in response: '${kw}'`);
      }
    }

    return {
      id: evalCase.id,
      passed: failures.length === 0,
      stop_reason: runResult.stop_reason,
      turns: runResult.turns.length,
      final_response: runResult.final_response,
      failures,
      duration_s: Math.round(duration * 1000) / 1000,
    };
  }
}
